File: py/泥视频.py
# -*- coding: utf-8 -*-
# 泥视频 - https://www.nivod.vip/
import re
import sys
import json
import time
from urllib.parse import quote, unquote
from pyquery import PyQuery as pq
sys.path.append('..')
from base.spider import Spider

class Spider(Spider):
    def init(self, extend=""):
        pass

    def getName(self):
        return "泥视频"

    def isVideoFormat(self, url):
        pass

    def manualVideoCheck(self):
        pass

    def destroy(self):
        pass

    host = 'https://www.nivod.vip'

    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
        'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        'Accept-Encoding': 'gzip, deflate, br',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1',
    }

    def homeContent(self, filter):
        """获取首页内容和分类"""
        try:
            response = self.fetch_with_encoding(self.host, headers=self.headers)
            doc = self.getpq(response.text)
            
            result = {}
            classes = []
            
            # 获取分类导航
            nav_items = doc('.navbar a')
            for item in nav_items.items():
                text = item.text().strip()
                href = item.attr('href')
                if text and href and href != '/' and '/t/' in href:
                    # 提取分类ID
                    type_id = href.split('/t/')[-1].rstrip('/')
                    if type_id.isdigit():
                        classes.append({
                            'type_name': text,
                            'type_id': type_id
                        })
            
            # 获取首页视频列表
            videos = []
            video_items = doc('.module-item')
            for item in video_items.items():
                try:
                    title = item.attr('title') or ''
                    href = item.attr('href') or ''
                    
                    if title and href:
                        # 提取视频ID
                        vod_id = href.split('/nivod/')[-1].rstrip('/')
                        
                        # 获取图片 - 优先获取data-original（真实图片），避免懒加载占位图
                        img_elem = item.find('img')
                        pic = ''
                        if img_elem:
                            # 优先获取data-original（真实图片URL），然后是data-src，最后是src
                            pic = img_elem.attr('data-original') or img_elem.attr('data-src') or img_elem.attr('src') or ''
                            if pic and not pic.startswith('http'):
                                pic = self.host + pic if pic.startswith('/') else ''
                        
                        # 获取备注信息
                        note_elem = item.find('.module-item-note')
                        remarks = note_elem.text() if note_elem else ''
                        
                        videos.append({
                            'vod_id': vod_id,
                            'vod_name': title,
                            'vod_pic': pic,
                            'vod_year': '',
                            'vod_remarks': remarks
                        })
                except Exception as e:
                    self.log(f"解析视频项时出错: {e}")
                    continue
            
            result['class'] = classes
            result['list'] = videos
            return result
            
        except Exception as e:
            self.log(f"获取首页内容时出错: {e}")
            return {'class': [], 'list': []}

    def homeVideoContent(self):
        """获取推荐视频"""
        return {'list': []}

    def categoryContent(self, tid, pg, filter, extend):
        """获取分类内容"""
        try:
            # 构建分类URL
            url = f"{self.host}/t/{tid}/"
            if int(pg) > 1:
                url = f"{self.host}/t/{tid}/page/{pg}/"
            
            response = self.fetch_with_encoding(url, headers=self.headers)
            doc = self.getpq(response.text)
            
            # 获取视频列表
            videos = []
            video_items = doc('.module-item')
            for item in video_items.items():
                try:
                    title = item.attr('title') or ''
                    href = item.attr('href') or ''
                    
                    if title and href:
                        # 提取视频ID
                        vod_id = href.split('/nivod/')[-1].rstrip('/')
                        
                        # 获取图片 - 优先获取data-original（真实图片），避免懒加载占位图
                        img_elem = item.find('img')
                        pic = ''
                        if img_elem:
                            # 优先获取data-original（真实图片URL），然后是data-src，最后是src
                            pic = img_elem.attr('data-original') or img_elem.attr('data-src') or img_elem.attr('src') or ''
                            if pic and not pic.startswith('http'):
                                pic = self.host + pic if pic.startswith('/') else ''
                        
                        # 获取备注信息
                        note_elem = item.find('.module-item-note')
                        remarks = note_elem.text() if note_elem else ''
                        
                        videos.append({
                            'vod_id': vod_id,
                            'vod_name': title,
                            'vod_pic': pic,
                            'vod_year': '',
                            'vod_remarks': remarks
                        })
                except Exception as e:
                    self.log(f"解析分类视频项时出错: {e}")
                    continue
            
            result = {
                'list': videos,
                'page': pg,
                'pagecount': 9999,  # 设置一个较大的值
                'limit': 80,
                'total': 999999
            }
            return result
            
        except Exception as e:
            self.log(f"获取分类内容时出错: {e}")
            return {'list': [], 'page': pg, 'pagecount': 1, 'limit': 80, 'total': 0}

    def detailContent(self, ids):
        """获取视频详情"""
        try:
            vod_id = ids[0]
            url = f"{self.host}/nivod/{vod_id}/"
            
            response = self.fetch_with_encoding(url, headers=self.headers)
            doc = self.getpq(response.text)
            
            # 获取标题
            title_elem = doc('h1')
            title = self.fix_encoding(title_elem.text()) if title_elem else ''

            # 获取视频信息
            info_elem = doc('.module-info')
            content = self.fix_encoding(info_elem.text()) if info_elem else ''
            
            # 获取播放源和播放列表
            play_from = []
            play_url = []
            
            # 获取播放源标签
            tab_items = doc('.module-tab-item')
            play_lists = doc('.module-play-list')
            
            for i, tab in enumerate(tab_items.items()):
                # 分别提取播放源名称和集数
                span_elem = tab.find('span')
                small_elem = tab.find('small')

                source_name = ''
                if span_elem:
                    source_name = self.fix_encoding(span_elem.text().strip())
                    # 如果有集数信息，添加到播放源名称后
                    if small_elem:
                        episode_count = self.fix_encoding(small_elem.text().strip())
                        source_name = f"{source_name}{episode_count}"
                else:
                    # 如果没有span元素，使用整个文本
                    source_name = self.fix_encoding(tab.text().strip())

                if source_name:
                    play_from.append(source_name)
                    
                    # 获取对应的播放列表
                    episodes = []
                    if i < len(play_lists):
                        episode_items = play_lists.eq(i).find('a')
                        for ep in episode_items.items():
                            ep_title = self.fix_encoding(ep.text().strip())
                            ep_href = ep.attr('href')
                            if ep_title and ep_href:
                                episodes.append(f"{ep_title}${ep_href}")
                    
                    play_url.append('#'.join(episodes))
            
            vod = {
                'vod_id': vod_id,
                'vod_name': title,
                'vod_pic': '',
                'vod_year': '',
                'vod_remarks': '',
                'vod_actor': '',
                'vod_director': '',
                'vod_content': content,
                'vod_play_from': '$$$'.join(play_from),
                'vod_play_url': '$$$'.join(play_url)
            }
            
            return {'list': [vod]}
            
        except Exception as e:
            self.log(f"获取视频详情时出错: {e}")
            return {'list': []}

    def searchContent(self, key, quick, pg="1"):
        """搜索内容"""
        try:
            # 使用正确的搜索URL格式
            search_url = f"{self.host}/s/-------------/"
            params = {'wd': key}

            response = self.fetch_with_encoding(search_url, params=params, headers=self.headers)
            doc = self.getpq(response.text)
            
            # 获取搜索结果
            videos = []
            video_items = doc('.module-item')
            for item in video_items.items():
                try:
                    # 搜索页面的结构不同，需要从内部链接获取信息
                    # 查找详情链接（通常是第一个或标题链接）
                    detail_links = item.find('a[href*="/nivod/"]')
                    if not detail_links:
                        continue

                    # 获取第一个详情链接
                    detail_link = detail_links.eq(0)
                    href = detail_link.attr('href')

                    if not href:
                        continue

                    # 提取视频ID
                    vod_id = href.split('/nivod/')[-1].rstrip('/')

                    # 获取标题 - 尝试多种方式
                    title = ''
                    # 方法1: 从链接的strong标签获取
                    strong_elem = detail_link.find('strong')
                    if strong_elem:
                        title = self.fix_encoding(strong_elem.text().strip())

                    # 方法2: 从图片的alt属性获取
                    if not title:
                        img_elem = item.find('img')
                        if img_elem:
                            title = self.fix_encoding(img_elem.attr('alt') or '')

                    # 方法3: 从链接文本获取
                    if not title:
                        title = self.fix_encoding(detail_link.text().strip())

                    if not title:
                        continue

                    # 获取图片 - 优先获取data-original（真实图片），避免懒加载占位图
                    img_elem = item.find('img')
                    pic = ''
                    if img_elem:
                        # 优先获取data-original（真实图片URL），然后是data-src，最后是src
                        pic = img_elem.attr('data-original') or img_elem.attr('data-src') or img_elem.attr('src') or ''
                        if pic and not pic.startswith('http'):
                            pic = self.host + pic if pic.startswith('/') else ''

                    # 获取备注信息
                    note_elem = item.find('.module-item-note')
                    remarks = self.fix_encoding(note_elem.text()) if note_elem else ''

                    videos.append({
                        'vod_id': vod_id,
                        'vod_name': title,
                        'vod_pic': pic,
                        'vod_year': '',
                        'vod_remarks': remarks
                    })
                except Exception as e:
                    self.log(f"解析搜索结果时出错: {e}")
                    continue
            
            return {'list': videos, 'page': pg}
            
        except Exception as e:
            self.log(f"搜索时出错: {e}")
            return {'list': [], 'page': pg}

    def playerContent(self, flag, id, vipFlags):
        """获取播放地址"""
        try:
            # 播放页面URL
            play_url = f"{self.host}{id}"
            
            response = self.fetch_with_encoding(play_url, headers=self.headers)
            doc = self.getpq(response.text)
            
            # 查找播放器配置
            scripts = doc('script')
            for script in scripts.items():
                script_text = script.text()
                if 'player' in script_text and ('url' in script_text):
                    # 尝试提取播放地址
                    url_match = re.search(r'"url"\s*:\s*"([^"]+)"', script_text)
                    if url_match:
                        video_url = url_match.group(1)
                        return {
                            'parse': 0,
                            'url': video_url,
                            'header': self.headers
                        }
            
            # 如果没有找到直接播放地址，返回播放页面让系统解析
            return {
                'parse': 1,
                'url': play_url,
                'header': self.headers
            }
            
        except Exception as e:
            self.log(f"获取播放地址时出错: {e}")
            return {
                'parse': 1,
                'url': f"{self.host}{id}",
                'header': self.headers
            }

    def localProxy(self, param):
        pass

    def fix_encoding(self, text):
        """修复UTF-8编码问题"""
        if not text:
            return text

        try:
            # 检查是否包含乱码特征（常见的UTF-8乱码模式）
            garbled_patterns = [
                '\u00e4\u00b8', '\u00e5', '\u00e6', '\u00e7', '\u00e8', '\u00e9',  # 常见乱码前缀
                '\u00c3\u00a4', '\u00c3\u00a5', '\u00c3\u00a6',  # UTF-8被误解为Latin1
                '\u00ef\u00bc', '\u00e2\u0080'  # 标点符号乱码
            ]

            has_garbled = any(pattern in text for pattern in garbled_patterns)

            if has_garbled:
                self.log("检测到编码问题，尝试修复...")

                # 方法1: 尝试Latin1->UTF-8转换
                try:
                    fixed = text.encode('latin1').decode('utf-8')
                    # 检查是否修复成功（包含中文字符）
                    if re.search(r'[\u4e00-\u9fff]', fixed):
                        self.log("使用Latin1->UTF-8修复成功")
                        return fixed
                except Exception as e:
                    self.log(f"Latin1->UTF-8修复失败: {e}")

                # 方法2: 尝试其他编码转换
                encodings = ['cp1252', 'iso-8859-1']
                for encoding in encodings:
                    try:
                        fixed = text.encode(encoding).decode('utf-8')
                        if re.search(r'[\u4e00-\u9fff]', fixed):
                            self.log(f"使用{encoding}->UTF-8修复成功")
                            return fixed
                    except:
                        continue

                self.log("编码修复失败，返回原文本")

            return text

        except Exception as e:
            self.log(f"编码修复异常: {e}")
            return text

    def fetch_with_encoding(self, url, **kwargs):
        """带编码处理的请求方法"""
        try:
            response = self.fetch(url, **kwargs)
            # 确保使用UTF-8编码
            response.encoding = 'utf-8'
            return response
        except Exception as e:
            self.log(f"请求失败: {e}")
            raise

    def getpq(self, text):
        """安全的pyquery解析"""
        try:
            return pq(text)
        except Exception as e:
            self.log(f"pyquery解析出错: {e}")
            try:
                return pq(text.encode('utf-8'))
            except:
                return pq('')

File: py/泥视频.js

var rule = {
    author: 'EylinSir修复版',
    title: '泥视频',
    类型: '影视',
    host: 'https://www.nivod.vip/',
    headers: {
        'User-Agent': 'MOBILE_UA'
    },
    编码: 'utf-8',
    timeout: 5000,
    homeUrl: '/',
    url: 'https://www.nivod.vip/k/fyfilter/',
    filter_url: '{{fl.cateId}}-{{fl.area}}-------fypage---{{fl.year}}',
    detailUrl: 'https://www.nivod.vip/nivod/fyid/',
    searchUrl: '/index.php/ajax/suggest?mid=1&wd=**&page=fypage&limit=30',
    搜索: 'json:list;name;pic;en;id',
    searchable: 1,
    quickSearch: 1,
    filterable: 1,
    limit: 10,
    double: false,
    class_name: '电影&电视剧&综艺&动漫',
    class_url: '1&2&3&4',

    filter_def: {
        1: { cateId: '1' },
        2: { cateId: '2' },
        3: { cateId: '3' },
        4: { cateId: '4' }
    },
    
    推荐: 'a:has(.lazyload);a&&title;.lazyload&&data-original;.module-item-note&&Text;a&&href',
    一级: 'a:has(.module-item-pic);a&&title;.lazyload&&data-original;.module-item-note&&Text;a&&href',

    二级: {
        title: 'h1&&Text',
        img: '.module-item-pic&&img&&data-src',
        desc: ';;.module-info-tag-link:eq(0)&&Text;.module-info-tag-link:eq(1)&&Text;.module-info-tag-link:eq(2)&&Text;.module-info-item:contains(集数)&&Text',
        content: '.module-info-introduction-content&&Text',
        tabs: '#y-playList&&span',
        lists: '.module-play-list:eq(#id)&&a'
    },

    sniffer: 0,
    isVideo: 'http((?!http).){12,}?\\.(m3u8|mp4|flv|avi|mkv|wmv|mpg|mpeg|mov|ts|3gp|rm|rmvb|asf|m4a|mp3|wma)',

    play_parse: true,
    lazy: `js:
try {
    let html = request(input);
    
    // 尝试多种匹配方式
    let match = null;
    let patterns = [
        /var player_.*?=\s*(\{.*?\})\s*</,
        /player_.*?=\s*(\{.*?\})\s*</,
        /var config\s*=\s*(\{.*?\})\s*</,
        /player_config\s*=\s*(\{.*?\})\s*</
    ];
    
    for (let pattern of patterns) {
        match = html.match(pattern);
        if (match && match[1]) {
            break;
        }
    }
    
    if (match && match[1]) {
        let kcode = JSON.parse(match[1]);
        let kurl = kcode.url;
        
        if (kcode.encrypt == '1') {
            kurl = unescape(kurl);
        } else if (kcode.encrypt == '2') {
            kurl = unescape(base64Decode(kurl));
        }
        
        if (kurl && /\.(m3u8|mp4|flv|avi|mkv)/.test(kurl)) {
            return {
                parse: 0,
                jx: 0,
                url: kurl,
                headers: {
                    'User-Agent': 'MOBILE_UA',
                    'Referer': 'https://www.nivod.vip/'
                }
            };
        }
    }
    
    // 如果上述方法失败，尝试直接提取iframe
    let iframeMatch = html.match(/<iframe[^>]*src=['"]([^'"]+)['"]/);
    if (iframeMatch && iframeMatch[1]) {
        return {
            parse: 1,
            jx: 0,
            url: iframeMatch[1]
        };
    }
    
} catch (e) {
    console.log('解析错误:', e.message);
}

// 默认返回
return {
    parse: 1,
    jx: 0,
    url: input
};
`,

    filter: {
        "1": [{
            "key": "cateId",
            "name": "类型",
            "value": [
                {"n": "全部", "v": "1"},
                {"n": "动作片", "v": "6"},
                {"n": "喜剧片", "v": "7"},
                {"n": "爱情片", "v": "8"},
                {"n": "科幻片", "v": "9"},
                {"n": "奇幻片", "v": "10"},
                {"n": "恐怖片", "v": "11"},
                {"n": "剧情片", "v": "12"},
                {"n": "战争片", "v": "20"},
                {"n": "纪录片", "v": "21"},
                {"n": "动画片", "v": "26"},
                {"n": "悬疑片", "v": "22"},
                {"n": "冒险片", "v": "23"},
                {"n": "犯罪片", "v": "24"},
                {"n": "惊悚片", "v": "45"},
                {"n": "歌舞片", "v": "46"},
                {"n": "灾难片", "v": "47"},
                {"n": "网络片", "v": "48"}
            ]
        }, {
            "key": "area",
            "name": "地区",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "大陆", "v": "大陆"},
                {"n": "香港", "v": "香港"},
                {"n": "台湾", "v": "台湾"},
                {"n": "美国", "v": "美国"},
                {"n": "欧美", "v": "欧美"},
                {"n": "日本", "v": "日本"},
                {"n": "韩国", "v": "韩国"},
                {"n": "泰国", "v": "泰国"},
                {"n": "其他", "v": "其他"}
            ]
        }, {
            "key": "year",
            "name": "年份",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "2025", "v": "2025"},
                {"n": "2024", "v": "2024"},
                {"n": "2023", "v": "2023"},
                {"n": "2022", "v": "2022"},
                {"n": "2021", "v": "2021"},
                {"n": "2020", "v": "2020"},
                {"n": "2019", "v": "2019"},
                {"n": "2018", "v": "2018"},
                {"n": "2017", "v": "2017"},
                {"n": "2016", "v": "2016"},
                {"n": "2015", "v": "2015"},
                {"n": "2014", "v": "2014"},
                {"n": "2013", "v": "2013"},
                {"n": "2012", "v": "2012"}
            ]
        }],
        "2": [{
            "key": "cateId",
            "name": "类型",
            "value": [
                {"n": "全部", "v": "2"},
                {"n": "国产剧", "v": "13"},
                {"n": "港台剧", "v": "14"},
                {"n": "日剧", "v": "15"},
                {"n": "韩剧", "v": "33"},
                {"n": "欧美剧", "v": "16"},
                {"n": "泰剧", "v": "34"},
                {"n": "新马剧", "v": "35"},
                {"n": "其他剧", "v": "25"}
            ]
        }, {
            "key": "area",
            "name": "地区",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "内地", "v": "内地"},
                {"n": "韩国", "v": "韩国"},
                {"n": "香港", "v": "香港"},
                {"n": "台湾", "v": "台湾"},
                {"n": "日本", "v": "日本"},
                {"n": "美国", "v": "美国"},
                {"n": "泰国", "v": "泰国"},
                {"n": "英国", "v": "英国"},
                {"n": "新加坡", "v": "Singapore"},
                {"n": "其他", "v": "其他"}
            ]
        }, {
            "key": "year",
            "name": "年份",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "2025", "v": "2025"},
                {"n": "2024", "v": "2024"},
                {"n": "2023", "v": "2023"},
                {"n": "2022", "v": "2022"},
                {"n": "2021", "v": "2021"},
                {"n": "2020", "v": "2020"},
                {"n": "2019", "v": "2019"},
                {"n": "2018", "v": "2018"},
                {"n": "2017", "v": "2017"},
                {"n": "2016", "v": "2016"},
                {"n": "2015", "v": "2015"},
                {"n": "2014", "v": "2014"},
                {"n": "2013", "v": "2013"},
                {"n": "2012", "v": "2012"}
            ]
        }],
        "3": [{
            "key": "cateId",
            "name": "类型",
            "value": [
                {"n": "全部", "v": "3"},
                {"n": "内地综艺", "v": "27"},
                {"n": "港台综艺", "v": "28"},
                {"n": "日本综艺", "v": "29"},
                {"n": "韩国综艺", "v": "36"},
                {"n": "欧美综艺", "v": "30"},
                {"n": "新马泰综艺", "v": "37"},
                {"n": "其他综艺", "v": "38"}
            ]
        }, {
            "key": "area",
            "name": "地区",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "内地", "v": "内地"},
                {"n": "港台", "v": "港台"},
                {"n": "日韩", "v": "日韩"},
                {"n": "欧美", "v": "欧美"}
            ]
        }, {
            "key": "year",
            "name": "年份",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "2025", "v": "2025"},
                {"n": "2024", "v": "2024"},
                {"n": "2023", "v": "2023"},
                {"n": "2022", "v": "2022"},
                {"n": "2021", "v": "2021"},
                {"n": "2020", "v": "2020"},
                {"n": "2019", "v": "2019"},
                {"n": "2018", "v": "2018"},
                {"n": "2017", "v": "2017"},
                {"n": "2016", "v": "2016"},
                {"n": "2015", "v": "2015"},
                {"n": "2014", "v": "2014"},
                {"n": "2013", "v": "2013"},
                {"n": "2012", "v": "2012"}
            ]
        }],
        "4": [{
            "key": "cateId",
            "name": "类型",
            "value": [
                {"n": "全部", "v": "4"},
                {"n": "国产动漫", "v": "31"},
                {"n": "日本动漫", "v": "32"},
                {"n": "韩国动漫", "v": "39"},
                {"n": "港台动漫", "v": "40"},
                {"n": "新马泰动漫", "v": "41"},
                {"n": "欧美动漫", "v": "42"},
                {"n": "其他动漫", "v": "43"}
            ]
        }, {
            "key": "area",
            "name": "地区",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "国产", "v": "国产"},
                {"n": "日本", "v": "日本"},
                {"n": "欧美", "v": "欧美"},
                {"n": "其他", "v": "其他"}
            ]
        }, {
            "key": "year",
            "name": "年份",
            "value": [
                {"n": "全部", "v": ""},
                {"n": "2025", "v": "2025"},
                {"n": "2024", "v": "2024"},
                {"n": "2023", "v": "2023"},
                {"n": "2022", "v": "2022"},
                {"n": "2021", "v": "2021"},
                {"n": "2020", "v": "2020"},
                {"n": "2019", "v": "2019"},
                {"n": "2018", "v": "2018"},
                {"n": "2017", "v": "2017"},
                {"n": "2016", "v": "2016"},
                {"n": "2015", "v": "2015"},
                {"n": "2014", "v": "2014"},
                {"n": "2013", "v": "2013"},
                {"n": "2012", "v": "2012"}
            ]
        }]
    }
}